### PR TITLE
NCF resume training

### PR DESCRIPTION
**Pull request by ForJadeForest**
## Description
This pr is for ncf tutorial of tf version to resume training with bigdl orca.
```
The first training results:
validation_loss: 0.2561998963356018
validation_accuracy: 0.8879268765449524
validation_auc: 0.9332888126373291
validation_precision: 0.7584264874458313
validation_recall: 0.6575862169265747

The resume training results: 
precision: 0.7698
recall: 0.6682
val_loss: 0.2453 
val_accuracy: 0.8929
val_auc: 0.9398
val_precision: 0.7580 
val_recall: 0.6952
```

### CODE REVIEW

File: python/orca/tutorial/NCF/tf_resume_spark_dataframe.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+# Step 3: Distributed training with Orca TF2 Estimator and load the model weight
+backend = 'ray'  # 'ray' or 'spark'
+est = Estimator.from_keras()
+est.load('NCF_model')
+
+batch_size = 10240
+train_steps = math.ceil(train_df.count() / batch_size)
+val_steps = math.ceil(test_df.count() / batch_size)
+callbacks = [tf.keras.callbacks.TensorBoard(log_dir="./log")]
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_spark_dataframe.py:47]: remove tensorboard in resume training?

File: python/orca/tutorial/NCF/tf_resume_spark_dataframe.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+est.fit(train_df,
+        epochs=2,
+        batch_size=batch_size,
+        feature_cols=feature_cols,
+        label_cols=label_cols,
+        steps_per_epoch=train_steps,
+        callbacks=callbacks)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_spark_dataframe.py:55]: add validation data in fit and remove evaluate?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 3]
```diff
@@ -152,6 +152,9 @@ def get_feature_cols():
     return feature_cols
 
 
+def get_label_col():
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:155]: get_label_cols

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 4]
```diff
@@ -152,6 +152,9 @@ def get_feature_cols():
     return feature_cols
 
 
+def get_label_cols():
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:155]: change Line142 to call this?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 5]
```diff
@@ -153,6 +153,10 @@ def get_feature_cols():
     return feature_cols
 
 
+def get_label_cols():
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:156]: same as above

File: python/orca/tutorial/NCF/tf_train_spark_dataframe.py [hunk 6]
```diff
@@ -68,12 +70,19 @@ def model_creator(config):
...
 batch_size = 10240
 train_steps = math.ceil(train_df.count() / batch_size)
 val_steps = math.ceil(test_df.count() / batch_size)
 callbacks = [tf.keras.callbacks.TensorBoard(log_dir=os.path.join(args.model_dir, "logs"))] \
     if args.tensorboard else []
 
+if args.scheduler:
+    lr_callback = tf.keras.callbacks.LearningRateScheduler(scheduler, verbose=1)
+    callbacks.append(lr_callback)
+    with open(os.path.join(args.model_dir, 'lr_callback.pkl'), 'wb')as f:
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_train_spark_dataframe.py:83]: need to check if model_dir is local or remote path. If remote, first save to tmp and put to remote. Check pytorch_dataloader to do similar things.

File: python/orca/tutorial/NCF/tf_train_spark_dataframe.py [hunk 7]
```diff
@@ -68,12 +70,19 @@ def model_creator(config):
                            backend=args.backend,
                            workers_per_node=args.workers_per_node)
 
+
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_train_spark_dataframe.py:73]: remove this

File: python/orca/tutorial/NCF/utils.py [hunk 8]
```diff
@@ -35,6 +35,8 @@ def parse_args(description):
                         help="Whether to use TensorBoard as the train callback.")
     parser.add_argument("--workers_per_node", type=int, default=1,
                         help="The number of workers on each node.")
+    parser.add_argument("--scheduler", action='store_true',
+                        help="Whether to use Scheduler as the train callback.")
```
Comment (hkvision) [inline python/orca/tutorial/NCF/utils.py:39]: Whether to use learning rate scheduler for training.

File: python/orca/tutorial/NCF/tf_train_xshards.py [hunk 9]
```diff
@@ -74,6 +76,12 @@ def model_creator(config):
 callbacks = [tf.keras.callbacks.TensorBoard(log_dir=os.path.join(args.model_dir, "logs"))] \
     if args.tensorboard else []
 
+if args.scheduler:
+    lr_callback = tf.keras.callbacks.LearningRateScheduler(scheduler, verbose=1)
+    callbacks.append(lr_callback)
+    with open(os.path.join(args.model_dir, 'lr_callback.pkl'), 'wb')as f:
+        pickle.dump(lr_callback, f)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_train_xshards.py:83]: save the scheduler before or after the training?

File: python/orca/tutorial/NCF/utils.py [hunk 10]
```diff
@@ -35,6 +35,8 @@ def parse_args(description):
                         help="Whether to use TensorBoard as the train callback.")
     parser.add_argument("--workers_per_node", type=int, default=1,
                         help="The number of workers on each node.")
+    parser.add_argument("--scheduler", action='store_true',
```
Comment (hkvision) [inline python/orca/tutorial/NCF/utils.py:38]: rename to lr_scheduler to avoid the same name with the scheduler function?

File: python/orca/tutorial/NCF/utils.py [hunk 11]
```diff
@@ -94,3 +96,10 @@ def init_orca(args, extra_python_lib=None):
              "'yarn-cluster', 'k8s-client', 'k8s-cluster', 'bigdl-submit' or 'spark-submit', "
              "but got " + args.cluster_mode)
     return sc
+
+
+def scheduler(epoch, lr):
+    if epoch < 1:
+        return lr
+    else:
+        return lr * 0.5
```
Comment (hkvision) [inline python/orca/tutorial/NCF/utils.py:105]: new line

File: python/orca/tutorial/NCF/tf_resume_xshard.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+# Step 2: Read and process data using Xshards
+train_data = XShards.load_pickle(os.path.join(args.data_dir, "train_processed_xshards"))
+test_data = XShards.load_pickle(os.path.join(args.data_dir, "test_processed_xshards"))
+feature_cols = get_feature_cols()
+label_cols = get_label_cols()
+
+
+# Step 3: Distributed training with Orca TF2 Estimator and load the model weight
+backend = 'ray'  # 'ray' or 'spark'
+est = Estimator.from_keras(backend=backend)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_xshard.py:44]: args.backend

File: python/orca/tutorial/NCF/process_xshards.py [hunk 13]
```diff
@@ -143,7 +143,7 @@ def rename(df, col):
     train_data, test_data = data.transform_shard(split_dataset).split()
 
     feature_cols = get_feature_cols()
-    label_cols = ["label"]
+    label_cols = get_label_cols()
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:146]: remove these two lines and directly call get_feature_cols() and get_label_cols() in return?

File: python/orca/tutorial/NCF/tf_resume_spark_dataframe.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+import tensorflow as tf
+
+from utils import *
+from process_spark_dataframe import get_feature_cols, get_label_cols
+
+from bigdl.orca.learn.tf2 import Estimator
+
+
+# Step 1: Init Orca Context
+args = parse_args("TensorFlow NCF Training with Spark DataFrame")
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_spark_dataframe.py:29]: Resume Training

File: python/orca/tutorial/NCF/tf_resume_xshard.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+from utils import *
+from process_xshards import get_feature_cols, get_label_cols
+
+from bigdl.orca.data import XShards
+from bigdl.orca.learn.tf2 import Estimator
+
+
+# Step 1: Init Orca Context
+args = parse_args("TensorFlow NCF Training with Spark DataFrame")
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_xshard.py:30]: wrong name

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 16]
```diff
@@ -138,8 +138,7 @@ def prepare_data(data_dir, neg_scale=4):
 
     occupation_num = df.agg({"occupation": "max"}).collect()[0]["max(occupation)"] + 1
     sparse_feats_input_dims.append(occupation_num)
-    feature_cols = get_feature_cols()
-    label_cols = ["label"]
+    feature_cols, label_cols = get_feature_label_cols()
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:141]: I mean in line 145, return  ...., get_feature_cols(), get_label_cols()

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 17]
```diff
@@ -148,8 +147,12 @@ def prepare_data(data_dir, neg_scale=4):
 
 
 def get_feature_cols():
-    feature_cols = ["user", "item"] + sparse_features + dense_features
-    return feature_cols
+    return ["user", "item"] + sparse_features + dense_features
+
+
+def get_feature_label_cols():
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:153]: just use get_label_cols()

File: python/orca/tutorial/NCF/tf_resume_spark_dataframe.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+callbacks = [tf.keras.callbacks.TensorBoard(log_dir=os.path.join(args.model_dir, "logs"))] \
+    if args.tensorboard else []
+
+if args.lr_scheduler:
+    lr_callback = tf.keras.callbacks.LearningRateScheduler(scheduler, verbose=1)
+    callbacks.append(lr_callback)
+
+est.fit(train_df,
+        epochs=5,
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_spark_dataframe.py:58]: maybe reduce some epochs?

File: python/orca/tutorial/NCF/tf_resume_spark_dataframe.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+        batch_size=batch_size,
+        feature_cols=feature_cols,
+        validation_data=test_df,
+        validation_steps=val_steps,
+        label_cols=label_cols,
+        steps_per_epoch=train_steps,
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_resume_spark_dataframe.py:64]: reorder the arguments: steps_per_epoch, validation_data, validation_steps, feature_cols, label_cols

**Comment by hkvision** [comment]
Paste some accuracy result before and after resume training?
